vignettes/backtesting.Rmd

```diff
@@ -0,0 +1,460 @@
+query and merged with `epix_merge()`:
+
+```{r, message = FALSE, warning = FALSE, eval = FALSE}
+library(epidatr)
+
+dv <- pub_covidcast(
+  source = "doctor-visits",
+  signals = "smoothed_adj_cli",
+  geo_type = "state",
+  time_type = "day",
+  geo_values = "ca,fl,ny,tx",
+  time_values = epirange(20200601, 20211201),
+  issues = epirange(20200601, 20211201)
+) %>%
+  rename(version = issue, percent_cli = value) %>%
+  as_epi_archive(compactify = TRUE)
+cases <- pub_covidcast(
+  source = "jhu-csse",
+  signals = "confirmed_7dav_incidence_prop",
+  geo_type = "state",
+  time_type = "day",
+  geo_values = "ca,fl,ny,tx",
+  time_values = epirange(20200601, 20211201),
+  issues = epirange(20200601, 20211201)
+) %>%
+  select(geo_value, time_value, version = issue, case_rate_7d_av = value) %>%
+  as_epi_archive(compactify = TRUE)
+archive_cases_dv_subset <- epix_merge(dv, y, sync = "locf", compactify = TRUE)
+```
+
+## Backtesting a simple autoregressive forecaster
+
+To start, let's use a simple autoregressive forecaster to predict the percentage of
+doctor's visits with CLI (`percent_cli`) in the future.
+
+The `arx_forecaster()` function wraps the autoregressive forecaster we need and
+comes with sensible defaults:
+
+- we specify the predicted outcome to be the percentage of doctor's visits with
+  CLI (`percent_cli`),
+- the predictors are the 7-day average of the case rate and the percentage of
+  doctor's visits with CLI (`percent_cli`),
+- we use a linear regression model as the engine,
+- the autoregressive features assume lags of 0, 7, and 14 days,
+- we forecast 7 days ahead.
+
+All these default settings and more can be seen by calling `arx_args_list()`:
+
+```{r}
+arx_args_list()
+```
+
+These can be modified as needed, by sending your desired arguments into
+`arx_forecaster(args_list = arx_args_list())`. For now we will use the defaults.
+
+__Note__: Unlike in the previous vignette, we will not train and forecast each
+geo location indivudally. Instead, we will use a __geo-pooled approach__, where
+we train the model on data from all states and territories combined. This is
+because the data is quite similar across states, and pooling the data can help
+improve the accuracy of the forecasts, while also reducing the susceptibility of
+the model to noise. In the interest of computational speed, we only use the 6
+state dataset here, but the full archive can be used in the same way and has
+performed well in the past. Implementation-wise, geo-pooling is achieved by
+simply dropping the `group_by(geo_value)` prior to `epix_slide()`.
+
+Let's use the `epix_as_of()` method to generate a snapshot of the archive at the
+last date, and then run the forecaster.
+
+```{r}
+# The .versions argument selects only the last version in the archive and
+# produces a forecast only on that date.
+forecasts <- archive_cases_dv_subset %>%
+  epix_slide(
+    ~ arx_forecaster(
+      .x,
+      outcome = "percent_cli",
```

Comment:
Do you really want `percent_cli` as the outcome? I think you want to predict cases, right?

Reply:
Yea, it's strange to forecast `percent_cli`, but it has the more drastic revision history. Cases from JHU barely have any revisions. This example could definitely be improved.

Reply:
This is a modification of the former `sliding.Rmd` right? I think it's important to have a believable forecast task. Predicting cli with something that almost certainly happens _after_ is not a great illustration.

Reply:
The previous `sliding.Rmd` seemed a bit unfocused to me: it seemed to both be trying to demonstrate version vs unversioned backtesting, but when plotting only compared the performance different engines for some reason. Like I said, forecasting cases make for a bad version vs unversioned example because there's no visible revisions, unlike percent_cli. 

The forecasts are not unbelievably good, you can see similar results in the new [epiprocess vignette](https://cmu-delphi.github.io/epiprocess/dev/articles/epi_archive.html#backtesting-a-simple-autoregressive-forecaster) using `prob_ar()`. How do you reason that we're getting future information for forecasting? We're using epix_slide, so all our data is versioned.

Reply:
percent_cli is constructed to be a leading indicator. We get that number before the person is tested and the test is reported to PH. So CLI precedes cases in time. Like predicting cases with deaths. The time arrow is backward. If you want to use cli as the outcome, we shouldn't use cases as a predictor.

Reply:
I see, makes sense. How about I just drop the exogenous variable for now and we can find one later?